Fixed returning correct length of dynamic shared memory buffer.

MoFtZ opened the pull request:
Fixes https://github.com/m4rs-mt/ILGPU/issues/356.

It appears that the `ArrayView` returned by `SharedMemory.GetDynamic<T>` currently returns -1 for its length on Cuda and OpenCL accelerators - the CPU accelerator is not affected.

On the Cuda accelerator, there is a [special register](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#special-registers-dynamic-smem-size) that contains the size of the dynamic memory (in bytes). This needs to be divided by the size of the data element type.

@m4rs-mt I'm not familiar with the OpenCL implementation of SharedMemory. How should this be handled?

m4rs-mt commented:
@MoFtZ Awesome work 🚀 As far as I understood the issue was that we are not aware of the actual length of a dynamic shared-memory allocation inside the compiler and your proposed IR value `DynamicSharedMemoryLengthValue` represents the length of a dynamic allocation. Should we change the value slightly to represent arbitrary dynamic-length arrays instead? I am asking because the function `CreateDynamicAllocaArray` receives an explicit address space and the `DynamicSharedMemoryLengthValue` seems to work for shared-memory allocations only. This can easily lead to programming issue in the future, I think 😟 

Regarding OpenCL: In OpenCL the amount of shared memory is specified via an explicit buffer input definition that receives the size of the buffer in bytes. I am currently not aware of any intrinsic functionality that should allow us to query this information. In the worst case, we have to pass an additional parameter to the kernel and store it in a "global" variable accessible by all functions.

MoFtZ commented:
Renamed `DynamicSharedMemoryLengthValue` to `DynamicMemoryLengthValue` as per review comments.

Added changes for OpenCL, passing the buffer length as an input parameter to the CLKernel, and then assigning it to a global variable.

MoFtZ commented on Src/ILGPU/Backends/OpenCL/CLInstructions.Data.cs:
```diff
@@ -157,6 +157,11 @@ partial class CLInstructions
         /// </summary>
         public const string GetLaneIndexOperation = "get_sub_group_local_id()";
 
+        /// <summary>
+        /// Resolves the current dynamic shared memory size.
+        /// </summary>
+        public const string GetDynamicSharedMemoryLengthOperation = "-1";
+
```
Obsolete - should be removed.

m4rs-mt commented on Src/ILGPU/Backends/OpenCL/CLCodeGenerator.Values.cs:
```diff
@@ -569,6 +569,52 @@ public void GenerateCode(SetField value)
+
+            // Resolve the name of the global variable containing the length of the
+            // shared dynamic memory buffer.
+            var dynamicView = value.GetFirstUseNode().ResolveAs<Alloca>();
```
Question: Currently, the invariant seems to be the case that the only occurrence of the `DynamicMemoryLengthValue` is the construction of an allocation. Is this guaranteed in all cases? Is there the possibility that we will add another pass in the future that could "rewire" the uses of this node to some other operation?

MoFtZ replied:
For now, it's guaranteed. If we use `DynamicMemoryLengthValue` somewhere else in the IR generation, then that could cause issues.

The problem is that `DynamicMemoryLengthValue` is used to construct the `Alloca`, so we cannot pass the ID of the IR node. If we could, then we could decouple the `CLCodeGenerator` from relying on `Alloca`.

m4rs-mt commented:
@MoFtZ Awesome patch 🚀  I added another comment regarding potential issues (I may be wrong here) that could arise in the future 🤓

m4rs-mt commented on Src/ILGPU/IR/Construction/IRBuilder.cs:
```diff
@@ -106,6 +106,22 @@ protected IRBuilder(BasicBlock basicBlock)
+        /// <param name="elementType">The element type.</param>
+        /// <param name="addressSpace">The target address space.</param>
+        /// <returns>A reference to the requested value.</returns>
+        public ValueReference CreateDynamicMemoryLengthValue(
```
Proposal: Maybe we should make this function `private` to the `IRBuilder` to ensure that it will be used in the scope of a dynamic allocation only 🤓 What do you think?

MoFtZ replied:
Fixed.
